cmd/mistryd/server.go

```diff
@@ -402,25 +406,121 @@ func (s *Server) ListenAndServe() error {
 	return s.srv.ListenAndServe()
 }
 
+// RebuildResult contains result data on the rebuild operation
+type RebuildResult struct {
+	successful  int
+	failed      int
+	pruneReport dockerTypes.ImagesPruneReport
+}
+
+func (r RebuildResult) String() string {
+	return fmt.Sprintf(
+		"Rebuilt: %d Failed: %d Pruned: %d Reclaimed: %d bytes",
+		r.successful, r.failed, len(r.pruneReport.ImagesDeleted), r.pruneReport.SpaceReclaimed)
+}
+
+// RebuildImages rebuilds images for all projects, and prunes any dangling images
+func RebuildImages(cfg *Config, log *log.Logger, out io.Writer, projects []string, failOnImageError bool) (RebuildResult, error) {
+	var err error
+	r := RebuildResult{}
+	if len(projects) == 0 {
+		projects, err = getProjects(cfg.ProjectsPath)
+		if err != nil {
+			return r, err
+		}
+	}
+
+	client, err := docker.NewEnvClient()
+	if err != nil {
+		return r, err
+	}
+
+	ctx := context.Background()
+	for _, project := range projects {
+		log.Printf("Rebuilding %s...\n", project)
+		j, err := NewJob(project, types.Params{}, "", cfg)
+		if err != nil {
+			r.failed++
+			log.Printf("Failed to instantiate %s job with error: %s\n", project, err)
+			if failOnImageError {
+				return r, err
+			}
+		} else {
+			err = j.BuildImage(ctx, cfg.UID, client, out, true, true)
+			if err != nil {
+				r.failed++
+				log.Printf("Failed to build %s job %s with error: %s\n", project, j.ID, err)
+				if failOnImageError {
+					return r, err
+				}
+			} else {
+				r.successful++
+			}
+		}
+	}
+	r.pruneReport, err = client.ImagesPrune(ctx, dockerFilters.Args{})
+	if err != nil {
+		return r, err
+	}
+	return r, nil
+}
+
+// PruneZombieBuilds removes any pending builds from the filesystem.
+func PruneZombieBuilds(cfg *Config) error {
+	projects, err := getProjects(cfg.ProjectsPath)
+	if err != nil {
+		return err
+	}
+	l := log.New(os.Stderr, "[cleanup] ", log.LstdFlags)
+
+	for _, p := range projects {
+		pendingPath := filepath.Join(cfg.BuildPath, p, "pending")
+		pendingBuilds, err := ioutil.ReadDir(pendingPath)
+		for _, pending := range pendingBuilds {
+			pendingBuildPath := filepath.Join(pendingPath, pending.Name())
+			err = cfg.FileSystem.Remove(pendingBuildPath)
+			if err != nil {
+				return fmt.Errorf("Error pruning zombie build '%s' of project '%s'", pending.Name(), p)
+			}
+			l.Printf("Pruned zombie build '%s' of project '%s'", pending.Name(), p)
+		}
+	}
+	return nil
+}
+
+func getProjects(path string) ([]string, error) {
+	folders, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+	projects := []string{}
```

Comment:
should use `make([]string, len(folders))` to build this slice

Reply:
What was the problem with just a simple slice without len? Why is the the trick mentioned in the link needed in our case?

Reply:
just a micro optimization

Reply:
not doing it